src/coreclr/vm/threads.h

```diff
@@ -3777,6 +3777,13 @@ class Thread
     // See ThreadStore::TriggerGCForDeadThreadsIfNecessary()
     bool m_fHasDeadThreadBeenConsideredForGCTrigger;
 
+    // lazily allocated
+    CLRRandom* m_pRandom;
+
+    public:
+        // TODO: where to delete the allocated CLRRandom object?
```

Comment:
I tried to find an example in the CLR code. The EventPipeThreadHolder* eventpipe_tls_instance is deleted when the free_thread_holder function is explicitly called. Same for t_pCurrentThreadLog that uses a different list mechanism but again, with a cleanup mechanism explicitly triggered.
I don't know when it is 100% sure that no code will allocate for a given thread. The allocation context is cleared in Thread::OnThreadTerminate. Do you confirm this could be the right place?

Reply:
No code should call CLRRandom during thread shutdown, so the ordering should not matter.

It can be something like this:
```cpp
//  .h file;

class CLRRandom
{
    ....

private:
    struct CLRRandomHolder
    {
        CLRRandom* _p;

        CLRRandomHolder()
        {
            _p = new CLRRandom();
            _p->Init();
        }

        ~CLRRandomHolder()
        {
            delete _p;
        }
    };

    static thread_local CLRRandomHolder t_instance;

public:
    CLRRandom* GetSharedRandom()
    {
        return t_instance._p;
    }
};

---
// .cpp file

thread_local CLRRandom::CLRRandomHolder CLRRandom::t_instance = CLRRandomHolder();
---

Reply:
✅ done